### PR TITLE
294 implement plotprecisionrecallcurve

### LINKED ISSUE #293: Implement PlotROC
## Issue description

The data analysis plot step `PlotROC` needs the plot method to be implemented.
This should probably be based around the [sklearn method](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.roc_curve.html) for plotting ROC.
For an example see [here](https://scikit-learn.org/1.1/auto_examples/model_selection/plot_roc.html)

## Acceptance Criteria
- [ ] Form accepting relevant parameters
- [ ] plot method that produces a ROC curve

### LINKED ISSUE #294: Implement PlotPrecisionRecallCurve
## Issue description

The `PlotPrecisionRecallCurve` needs a plot_method. See the [sklearn implementation](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.precision_recall_curve.html). For some background, have a look at this [blog post](https://medium.com/@douglaspsteen/precision-recall-curves-d32e5b290248)

## Acceptance Criteria
- [ ] Form accepting relevant input parameters
- [ ] plot method producing a precision recall curve

### PR DESCRIPTION
**Pull request by tE3m**
## Description
fixes #293 and #294 

Introduces new plot methods `roc_plot` and `precision_recall_plot` that use plotly (cf [this guide](https://plotly.com/python/roc-and-pr-curves/) instead of the previous methods that used scikit's built-in plot methods that rely on matplotlib. Also fixes some aspects of the classification steps (svm, randomforest).

## Changes

- `backend/protzilla/data_analysis/plots.py` - implementation of the methods
- `backend/protzilla/methods/data_analysis.py` - adapt the steps to use the new methods
- `backend/protzilla/data_analysis/classification.py` - adapt output format to avoid unnecessary conversions
- `backend/protzilla/steps.py` 
  - avoid overwriting default values if not present in inputs
  -  temporarily change `insert_dataframes` (otherwise model-outputs don't work, should be done properly by the PR that introduces serialization of models)

## Testing
1. Import protein and metadata
2. do some preprocessing (e.g. the standard workflow up to Normalisation: Median)
3. perform classification through either SVM or Random Forest
4. connect equally named handles to ROC or Precision Recall steps
5. plot
6. profit?

## PR checklist
**Development**
- [ ] If necessary, I have updated the documentation (README, docstrings, etc.)
- [x] If necessary, I have created / updated tests.
 
**Mergeability**
- [ ] main-branch has been merged into local branch to resolve conflicts - no: not mergable until models are properly supported as outputs
- [x] The tests and linter have passed AFTER local merge (except for those known to be failing)
- [x] The backend code has been formatted with `black`
 
**Code review**
- [x] I have self-reviewed my code.
- [ ] At least one other developer reviewed and approved the changes

### CODE REVIEW

File: backend/protzilla/data_analysis/model_evaluation_plots.py [hunk 1]
```diff
@@ -6,6 +6,8 @@
 from backend.protzilla.utilities.utilities import fig_to_base64
 
 
+# TODO: might be useful for validation but incompatible with frontend,
+# since the builtin plot method returns a matplotlib object
```
Comment (NeleRiediger) [inline backend/protzilla/data_analysis/model_evaluation_plots.py:10]: Don't know if I'm getting this comment wrong, but is it supposed to stay there?

File: backend/protzilla/data_analysis/model_evaluation_plots.py [hunk 2]
```diff
@@ -22,17 +24,17 @@ def precision_recall_curve_plot(model, input_test_df, labels_test_df, plot_title
...
-        model, input_test_df, labels_test_df["Encoded Label"]
+        model, input_test_df, labels_test_df
     )
     display.plot(color=PLOT_PRIMARY_COLOR)
     plot.title(plot_title)
     return [fig_to_base64(display.figure_)]
 
 
+# TODO: might be useful for validation but incompatible with frontend,
+# since the builtin plot method returns a matplotlib object
```
Comment (NeleRiediger) [inline backend/protzilla/data_analysis/model_evaluation_plots.py:37]: Here as well?

File: backend/protzilla/data_analysis/model_evaluation_plots.py [hunk 3]
```diff
@@ -49,12 +51,8 @@ def roc_curve_plot(model, input_test_df, labels_test_df, plot_title=None):
...
-    input_test_df = input_test_df.set_index("Sample")
-    _, labels_test_df = encode_labels(labels_test_df, "Label")
 
-    display = RocCurveDisplay.from_estimator(
-        model, input_test_df, labels_test_df["Encoded Label"]
-    )
+    display = RocCurveDisplay.from_estimator(model, input_test_df, labels_test_df)
     display.plot(color=PLOT_PRIMARY_COLOR)
     plot.title(plot_title)
     return [fig_to_base64(display.figure_)]
```
Comment (jorisfu) [inline backend/protzilla/data_analysis/model_evaluation_plots.py:58]: If we really don't need these anymore, I suggest removing them. Alternatively we could also display them now.

File: backend/protzilla/methods/data_analysis.py [hunk 4]
```diff
@@ -1681,6 +1695,9 @@ class ClassificationSVM(ClassificationStep):
         "y_test_df",
     ]
 
+    # TODO: should either be set via form_inputs or removed from the method's parameters
+    internal_inputs = {"max_iter", "coef0", "gamma", "class_weight", "probability"}
+
```
Comment (jorisfu) [inline backend/protzilla/methods/data_analysis.py:1700]: You still plan on adding those?

**Review by NeleRiediger — APPROVED** [review]
Testing worked as expected. Take a quick look at my comments, but generally I think everything should be fine.

File: backend/protzilla/data_analysis/model_evaluation_plots.py (hunk 1 above)
Comment (tE3m) [reply backend/protzilla/data_analysis/model_evaluation_plots.py:10]: yes, since I've implemented the plot methods "myself" I thought it was worth to keep them to be able to compare if the new plots are doing the correct thing. Coincidentally, Joris added the ability for PNG plots in a different branch, so they could even be displayed to users now

File: backend/protzilla/steps.py [hunk 5]
```diff
@@ -220,7 +220,13 @@ def insert_dataframes(self, steps: StepManager) -> None:
...
                     f"Step {source} has no output with key {source_handle}, but was set to be the input in {target} for key {target_handle}"
                 )
-            self.inputs[target_handle] = source_output.copy()
+            # TODO: temporary measure while support for model outputs isn't properly finished
+            # the model instances don't implement a .copy() method
+            self.inputs[target_handle] = (
+                source_output.copy()
+                if isinstance(source_output, (pd.DataFrame, list))
+                else source_output
+            )
```
Comment (jorisfu) [inline backend/protzilla/steps.py:230]: Cleaner would be to check `if hasattr(source_output, 'copy') and callable(getattr(source_output, 'copy'))` to see if we implement the copy trait

**Comment by jorisfu** [comment]
Whoops I'm late to the party, should've reviewed before lunch

File: backend/protzilla/methods/data_analysis.py (hunk 4 above)
Comment (tE3m) [reply backend/protzilla/methods/data_analysis.py:1700]: no

File: backend/protzilla/steps.py (hunk 5 above)
Comment (tE3m) [reply backend/protzilla/steps.py:230]: looks good, can you maybe add this on the 179 branch directly?
